File: code-proof-paper/code/src/proof/spec.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol, Any


class Property(Protocol):
    """Protocol for a property to be proven by code-backed checks."""

    def check(self, candidate: Any) -> bool:  # pragma: no cover - trivial protocol
        ...


@dataclass(frozen=True)
class ProofResult:
    name: str
    passed: bool
    message: str = ""


def verify_property(property_name: str, property_under_test: Property, generator: Any, num_trials: int = 1000) -> ProofResult:
    """
    Execute property checks over generated candidates.
    This is a placeholder scaffold; replace generator with your domain generator and
    tighten the test to your proof obligation.
    """
    passed_all = True
    for _ in range(max(1, num_trials)):
        candidate = next(generator)
        if not property_under_test.check(candidate):
            passed_all = False
            break
    return ProofResult(name=property_name, passed=passed_all)

File: implementation/src/proof/spec.ts
export interface Property<T = unknown> {
  check(candidate: T): boolean
}

export type GeneratorFn<T> = () => T

export interface ProofResult {
  name: string
  passed: boolean
  message?: string
}

export function verifyProperty<T>(
  propertyName: string,
  propertyUnderTest: Property<T>,
  generator: GeneratorFn<T>,
  numTrials = 1000
): ProofResult {
  let passed = true
  for (let i = 0; i < Math.max(1, numTrials); i++) {
    const candidate = generator()
    if (!propertyUnderTest.check(candidate)) {
      passed = false
      break
    }
  }
  return { name: propertyName, passed }
}

